### PR TITLE
fix: Added null check for partitions batches

### PR DESCRIPTION
**Pull request by lukas-gust**
### Description

This is my first contribution, thank you in advance.

There is an issue when the batch partition values take on a null value and are of certain types. When filtering source rows based on partition if an integer partition value show null the current behavior is to convert that to a string in the jinja macro resulting in `partition_key=None` and for normal values `partition_key=1234`. The former is incorrect syntax and behavior.

The aim of this is to fix that issue by evaluating if the partition value is none and setting the filter comparison function to something like `partition_key is null`.

## Models used to test - Optional
```
{# This table is merged on load (falls back to append currently). Needs deduplication. #}
{# This is an iceberg table type from the project config #}

{{ config(
    materialized = 'incremental',
    incremental_strategy = 'merge',
    partitioned_by = ['status', 'organization_id', 'hour(created_at)'],
    unique_key = ['id'],
) }}

with using_tickets as (

    select
        id,
        "url",
        tags,
        via__channel as via_channel,
        json_parse("via__source__from") as via_source_from,
        via__source__to as via_source_to,
        via__source__rel as via_source_rel,
        cast(created_at as timestamp(6)) as created_at,
        cast(updated_at as timestamp(6)) as updated_at,
        "type",
        "subject",
        raw_subject,
        "description",
        priority,
        "status",
        requester_id,
        submitter_id,
        assignee_id,
        group_id,
        has_incidents,
        is_public,
        custom_status_id,
        ticket_form_id,
        brand_id,
        allow_channelback,
        allow_attachments,
        from_messaging_channel,
        cast(generated_timestamp as timestamp(6)) as generated_timestamp,
        channel,
        property,
        ticket_type,
        organization_id,
        opt_out_reason,
        internal_escalation,
        rsa_issue_resolution,
        escalation_type,
        origin,
        "_dlt_id" as dlt_id,
        "_dlt_load_id" as dlt_load_id
    from
        {{ source(
            'bronze_zendesk_support',
            'tickets'
        ) }}

    {% if is_incremental() %}
    where updated_at >= coalesce((select max(updated_at) from {{ this }}), date('1900-01-01'))
    {% endif %}
),

using_extracted_json as (
    select
        {# Some json extraction logic #}
    from
        using_tickets
),

updates as (
    select
        *
        row_number() over (partition by id order by updated_at desc) as row_number
    from
        using_extracted_json
    
    {% if is_incremental() %}
    where
        id in (select id from {{ this }})
    {% endif %}
),

inserts as (
    select
        *
        row_number() over (partition by id order by updated_at desc) as row_number
    from
        using_extracted_json
    
    {% if is_incremental() %}
    where
        id not in (select id from {{ this }})
    {% endif %}
)

select 
    *
from
(
    select * from updates 
    union 
    select * from inserts
) as results
where row_number = 1
```

## Checklist
- [x] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [x] You kept your Pull Request small and focused on a single feature or bug fix.
- [x] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

**Comment by nicor88** [comment]
@lukas-gust looks good - thanks for the fix. 
We should consider to cover this edge case with some functional testing: https://github.com/dbt-athena/dbt-athena/blob/main/tests/functional/adapter/test_partitions.py - we could add something here. WDYT? - we can create for example a dataset with more than 100 dates and then append some null values.

**Comment by lukas-gust** [comment]
The new test I added fails on the record count. I need to step away for the time being.

**Comment by nicor88** [comment]
@lukas-gust it's quite weird, the record count is 3 instead of 212...I will have a deep look into it. Maybe we are hitting another edge case, but I can't figure it out at the moment.

File: tests/functional/adapter/test_partitions.py [hunk 1]
```diff
@@ -17,6 +17,24 @@
 cross join unnest(date_array) as t2(date_column)
 """
 
+test_null_valued_partitions_model_sql = """
```
Comment (nicor88) [inline tests/functional/adapter/test_partitions.py:20]: @lukas-gust I changed the test a bit to cover a simple case

**Comment by nicor88** [comment]
@lukas-gust I modified a bit the test case to cover a situation where we have only 1 partition column (int) that contains NULL values - should pass.

Anyhow I created another test case:

```
with data as (
    select
        cast(random() * 100 as int) col_1,
        row_number() over() as id
    from
        unnest(sequence(1, 200))
)

select
    'test case sequence' as record_type,
    id,
    case when col_1 < 50 then null else col_1 end as col_1
from data
union all
select
    'test case set 1' as record_type,
    NULL as id,
    NULL as col_1
union all
select
    'test case set 2' as record_type,
    NULL as id,
    23 as col_1
union all
select
    'test case set 3' as record_type,
    50 as id,
    NULL as col_1
```

somehow that does work - I expect 202 columns - but I get only a part of those.

The issue seems to be due to the fact that with NULLs the `where` generated statements doesn't cover all cases.

```
where (id=200 and col_1 is null) or (id is null and col_1=23) or (id is null and col_1 is null)
```

or 

```
where (id=100 and col_1 is null) or (id=101 and col_1=57) or (id=102 and col_1=97) or (id=103 and col_1=67) or (id=104 and col_1 is null) or (id=105 and col_1 is null) or (id=106 and col_1 is null) or (id=107 and col_1 is null) or (id=108 and col_1=84) or (id=109 and col_1=83) or (id=110 and col_1 is null) or (id=111 and col_1 is null) or (id=112 and col_1 is null) or (id=113 and col_1=96) or (id=114 and col_1=64) or (id=115 and col_1=52) or (id=116 and col_1 is null) or (id=117 and col_1 is null) or (id=118 and col_1 is null) or (id=119 and col_1=91) or (id=120 and col_1=80) or (id=121 and col_1 is null) or (id=122 and col_1 is null) or (id=123 and col_1=76) or (id=124 and col_1 is null) or (id=125 and col_1=65) or (id=126 and col_1 is null) or (id=127 and col_1=51) or (id=128 and col_1=89) or (id=129 and col_1 is null) or (id=130 and col_1=89) or (id=131 and col_1=98) or (id=132 and col_1 is null) or (id=133 and col_1=68) or (id=134 and col_1 is null) or (id=135 and col_1 is null) or (id=136 and col_1=96) or (id=137 and col_1=73) or (id=138 and col_1=92) or (id=139 and col_1 is null) or (id=140 and col_1=93) or (id=141 and col_1 is null) or (id=142 and col_1=53) or (id=143 and col_1=86) or (id=144 and col_1=51) or (id=145 and col_1 is null) or (id=146 and col_1 is null) or (id=147 and col_1 is null) or (id=148 and col_1=89) or (id=149 and col_1=55) or (id=150 and col_1 is null) or (id=151 and col_1 is null) or (id=152 and col_1 is null) or (id=153 and col_1 is null) or (id=154 and col_1 is null) or (id=155 and col_1=72) or (id=156 and col_1=95) or (id=157 and col_1=88) or (id=158 and col_1 is null) or (id=159 and col_1=55) or (id=160 and col_1 is null) or (id=161 and col_1 is null) or (id=162 and col_1 is null) or (id=163 and col_1 is null) or (id=164 and col_1=53) or (id=165 and col_1 is null) or (id=166 and col_1 is null) or (id=167 and col_1=63) or (id=168 and col_1 is null) or (id=169 and col_1=90) or (id=170 and col_1=60) or (id=171 and col_1 is null) or (id=172 and col_1=96) or (id=173 and col_1=66) or (id=174 and col_1 is null) or (id=175 and col_1 is null) or (id=176 and col_1 is null) or (id=177 and col_1 is null) or (id=178 and col_1=66) or (id=179 and col_1 is null) or (id=180 and col_1=92) or (id=181 and col_1 is null) or (id=182 and col_1=55) or (id=183 and col_1 is null) or (id=184 and col_1 is null) or (id=185 and col_1 is null) or (id=186 and col_1 is null) or (id=187 and col_1=50) or (id=188 and col_1=66) or (id=189 and col_1 is null) or (id=190 and col_1 is null) or (id=191 and col_1 is null) or (id=192 and col_1 is null) or (id=193 and col_1 is null) or (id=194 and col_1=56) or (id=195 and col_1 is null) or (id=196 and col_1=98) or (id=197 and col_1=91) or (id=198 and col_1 is null) or (id=199 and col_1 is null)
```
That means that your proposed change works doesn't work in all the scenarions.

**Comment by nicor88** [comment]
I just tested locally - let's add another test case
```
with data as (
    select
        row_number() over() as id,
        row_number() over() + 200 as col_1
    from
        unnest(sequence(1, 200))
)

select
    'test case sequence' as record_type,
    case when id < 50 then null else id end as id,
    case when col_1 > 150 then null else col_1 end as col_1
from data
union all
select
    'test case set 1' as record_type,
    NULL as id,
    NULL as col_1
union all
select
    'test case set 2' as record_type,
    2 as id,
    NULL as col_1
union all
select
    'test case set 3' as record_type,
    NULL as id,
    23 as col_1

```

where we partition by id and col_1 - should work.

We expect 203 records.

Also as the count of null is deterministic we can say for example that we expect 51 records with NULL id.

And also to confirm your fix work - the issue was just in the test case dataset in combination with random.

**Comment by lukas-gust** [comment]
Thank you for the help! That's interesting, I mean it makes sense, but I'm trying to think if that's an issue. So each time we iterate we generate a new set of random data so we can't know how many records there were. Is this  side effect of the tests or would this occurr naturally of my query was structured the way I had it?

**Comment by nicor88** [comment]
The issue is that random() gets evaluate on each query where it's called.

Imaging that in your dataset you have id=345 (generate by random) - but when we do select distinct id from (....sql query with the dataset) -> random() will return another value that is not 345, that create a non deterministic behavior - that is different in real case datasets.

**Comment by lukas-gust** [comment]
That makes sense. Wouldn't it be the same situation for something like current_timestamp? Basically wondering if we ought to stage a non-partitioned temp table to read from instead. That way this isn't possible. Or is it a desired behavior. Thoughts? Glad we were able to work through the test.

**Comment by nicor88** [comment]
I believe that persisting the SQL of the model to a not-partitioned table is indeed better - not sure why @svdimchenko didn't go for that - but we can certainly add it. I will make the SQL queries that are run in case we hit partition limitations easier to debug - but that can be done in another PR.

**Comment by lukas-gust** [comment]
100% agree. @nicor88  Yes I will work on it later tonight.

**Comment by svdimchenko** [comment]
> I believe that persisting the SQL of the model to a not-partitioned table is indeed better - not sure why @svdimchenko didn't go for that

Hi there! @nicor88 It's dual question (if I understand it correctly). From one point of view it will be more easy to debug. On the other side, once we create not partitioned tmp table we'll need to full scan this tmp table on every inseert iteration into target relation. This approach will lead to costs growth I guess. But we can discuss that of course.

**Comment by nicor88** [comment]
@svdimchenko  
>  thought once we create not partitioned tmp table we'll need to full scan this tmp table on every insert iteration into target relation. This approach may lead to costs growth I guess. But we can discuss that of course.
However not I see that we'll run model's sql consecutively, which may be even more costly and consume more resources. My bad and it will be great enhancement to rework tmp relation creation.

Few reasons why to use a tmp table (using parquet):
* we force the usage of a columnar format - (imagine cases where the user use JSON) and we run many queries for the insert statements in this not optimal source table
* the distinct queries to get partitions will be easier to read - `select my_partition_column from my_tmp_table`
* the insert queries will be also easier to read - `insert into my_final_table(x, y, z) select x,y,z from my_tmp_table where z='a' and z='c'`
* in case something goes wrong we have my_tmp_table that we could inspect 

 
Notes
* even using a tmp table as parquet we cannot avoid full scans on it - for that we need to use partitions
* I don't think that using a tmp table will create much more costs - there will be only 1 extra first query where we persist the dataset, but the consecutive queries on the tmp could be less costly in term of data scans as we force usage of parquet - and scan only on columns that we select.
* persisint the sql query in a tmp table will improve scenarios where random or current timestamp functions are used, creating a more deterministic behavior

Happy to discuss more about it :)

**Review by nicor88 — APPROVED** [review]

**Comment by nicor88** [comment]
@svdimchenko do you want to look a this as well?

**Review by svdimchenko — APPROVED** [review]

**Comment by nicor88** [comment]
great work @lukas-gust, and nice first contribution 💯
